### PR TITLE
Adds CMEMS/ETOPO/WOA environment variables to a dataset

### PR DESCRIPTION
**Pull request by jeremyfix**
This PR is a galaxy tool wrapper around the [medenv](https://github.com/jeremyfix/medenv/) I developed which allows to request the etopo, cmems and woa products to add environmental variables to a dataset.

The tool is hopefully generic in the sense you can customize, within the galaxy, the column names for the longitude, latitude, time and depth of your requests. 

It allows to add 21 variables. Bathymetry is brought by etopo and there others by the CMEMS mediterranean products. The medenv library is flexible enough so that we can latter add more cmems products. Although medenv does have a tool to request woa and request some variables, it is not yet availavble from the high level api of medenv nor from the galaxy tool. 

When requesting variables for a specific latitude, longitude, depth, time, you can indicate a margin around the latitude/longitude. medenv will collect the variables in this longitude_range x latitude_range at the specific depth/time and merge (average) these values. 

**Important** : One of the limitations is that you must explicitely write your CMEMS credentials within the galaxy tool. Which is certainly bad because of possible leaks of these credentiels but I do not see how to protect this potential leak in the way galaxy tools are defined. 

The test included in the PR is also only testing for the bathymetry request, targeting the public etopo. And there is no tests for the cmems requests that would require credentials to be provided.

### CODE REVIEW

**Comment by yvanlebras** [comment]
THANK YOU Jeremy for this contribution! We will go forward on the review and first of all, I need to check if we can at least use a token or something like this for cmems tools.

**Comment by yvanlebras** [comment]
Going forward, I am looking at CMEMS data access https://data.marine.copernicus.eu/product/GLOBAL_ANALYSIS_FORECAST_BIO_001_028/services . Maybe it can be of interest to have your review and help on this @Marie59 as it seems to me this can be a good tool for FAIR EASE communities?

**Comment by bgruening** [comment]
On EU, we do have a Copernicus key and using it for this tool: https://usegalaxy.eu/root?tool_id=toolshed.g2.bx.psu.edu/repos/climate/c3s/c3s/0.1.0

Maybe you can take inspiration from that tool or even extend this tool?

ping @annefou 

See especially https://github.com/NordicESMhub/galaxy-tools/tree/master/tools/essential_climate_variables

And this config file: https://github.com/NordicESMhub/galaxy-tools/blob/master/tools/essential_climate_variables/cdsapirc.cfg

**Comment by yvanlebras** [comment]
it seems very related! THANK YOU Björn! So we can look to create a copernicus CMEMS "usegalaxy.eu account" and store credential on the Galaxy Europe server backoffice if I understand well.

**Comment by annefou** [comment]
Just as a side note: in the tool  https://github.com/NordicESMhub/galaxy-tools/tree/master/tools/essential_climate_variables we have a key that can be used by everyone because the data license authorises us to do so but for https://github.com/NordicESMhub/galaxy-tools/tree/master/tools/c3s each user needs to register and add their own credentials (which is expected for some data, depending on the license).

**Comment by yvanlebras** [comment]
Crystal clear Jeremy !

I think maybe @Marie59 can investigate this particular credential issue and then see with @bgruening how it is better to implement in Galaxy ! Then, I need to check how we can deal with that on our "French Galaxy AI" instance. I will come back to you Jeremy!

**Comment by Marie59** [comment]
Yes, this something we already talked about for the credentials. I think we can add another tool to retrieve CMEMS data on galaxy and this tool too would really be of interest for me.

**Comment by bgruening** [comment]
@jeremyfix the credentials depends one what is allowed. In Galaxy, as Anne above pointed out, we can have global credentials. Those instance-global credentials needs to be setup by a Galaxy admin. And then all users of this instance will use them - probably not even knowing they are using them. However, is is only possible if Copernicus allows that use-case.

If this is not possible. If every user needs to have their own credentials a user can store them in their user-preferences and tool can access per-user-credentials. There is no file upload needed. 

This is what Anne showed here: https://github.com/NordicESMhub/galaxy-tools/blob/master/tools/c3s/c3s.xml#L12 <- its access user-preferences.

**Comment by bgruening** [comment]
Long term it would be nice if we could get a connection to some higher Copernicus dev or PI. Galaxy could do much more in terms of hand-shaking with Copernicus. We could work towards a OIDC-based authentification between both services or push data from Copernicus to Galaxy. All that is possible but needs some work on the Copernicus side as well.

**Comment by bgruening** [comment]
@jeremyfix @yvanlebras would it be better to contribute this tool to the climate repo? So that we have all Copernicus tools together?

File: tools/medenv/med_environ.xml [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+        From the input dataset, it will request copernicus for environment variables.
+    </description>
+
+	<requirements>
+        <requirement type="package" version="2.1.3">pandas</requirement>
+        <requirement type="package" version="4.9.3">lxml</requirement>
+        <requirement type="package" version="1.0.4">medenv</requirement>
+	</requirements>
+
+    <command>
```
Comment (bgruening) [inline tools/medenv/med_environ.xml:12]: feel free to add https://docs.galaxyproject.org/en/latest/dev/schema.html#tool-creator-person

File: tools/medenv/med_environ.xml [hunk 2]
```diff
@@ -0,0 +1,129 @@
+<tool id="iabiodiv_smartbiodiv_med_environ" name="Adds the environment variables to a dataset" version="0.0.1">
```
Comment (bgruening) [inline tools/medenv/med_environ.xml:1]: please add a profile
```suggestion
<tool id="iabiodiv_smartbiodiv_med_environ" name="Adds the environment variables to a dataset" version="0.0.1" prfile="22.05">
```

File: tools/medenv/med_environ.xml [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+        <param name="tol_spatial_coordinates" type="float" value="0.2" label="A tolerance on the latitude/longitude for the request point" optional="False"/>
+		<param name="depth_key" type="text" value="Depth" label="The key of the column for the observation time" optional="False"/>
+		<param name="date_key" type="text" value="SurveyDate" label="The key of the column for the observation time" optional="False"/>
+		<param name="cmems_username" type="text" value="xxxxx" label="Your login to access CMEMS data"/>
+        <param name="cmems_password" type="text" value="xxxxx" label="Your password to access CMEMS data"/>
+        <param name="verbose" type="boolean" value="False" label="Verbose in the standard output of the tool"/>
+    </inputs>
+
+    <outputs>
+        <data format="csv" name="out_file" metadata_source="input" label="Dataset augmented with the environmental variables"/>
```
Comment (bgruening) [inline tools/medenv/med_environ.xml:53]: Can we output tab seperated files? `tabular` in the Galaxy ontoogy?

File: tools/medenv/med_environ.xml [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+
+            - "mixed-layer-thickness",
+            - "sea-surface-temperature",
+            - "sea-surface-salinity",
+            - "sea-surface-above-geoid",
+            - "surface-partial-pressure-co2",
+            - "surface-co2-flux".
+
+		</help>
+    <citations>
```
Comment (bgruening) [inline tools/medenv/med_environ.xml:125]: you can use type=bibtex here and cite your github repo if you like

**Review by bgruening — COMMENTED** [review]
Please use 4 space indentation.

File: tools/medenv/med_environ.xml (hunk 2 above)
Comment (jeremyfix) [reply tools/medenv/med_environ.xml]: I added the profile version you suggested although did not review all the requirements that this profile make.

File: tools/medenv/med_environ.xml (hunk 4 above)
Comment (jeremyfix) [reply tools/medenv/med_environ.xml:129]: Done as well ; Although not cited yet assigned a DOI :)

**Comment by jeremyfix** [comment]
@bgruening  I should have fixed all your requests.

File: tools/medenv/med_environ.xml [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+python '$__tool_directory__/med_environ.py'
+    --datafile '${input}'
+    --out_file '${out_file}'
+    --lat_key  '${lat_key}'
+    --long_key '${long_key}'
+    --date_key '${date_key}'
+    --depth_key '${depth_key}'
+    --variables '${env_variables}'
+    --tol_spatial_coordinates '${tol_spatial_coordinates}'
+    --cmems_username '${cmems_username}'
```
Comment (bgruening) [inline tools/medenv/med_environ.xml:23]: This should not come from the tool form. Tool parameters are stored for reproducibility and are on purpose exposed to the world - FAIR.

Please take those from the ENV variables from the shell. As shown in @annefou tools.

File: tools/medenv/med_environ.xml [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+            <option value="surface-co2-flux">surface CO2 flux</option>
+            <option value="mixed-layer-thickness">mixed layer thickness</option>
+        </param>
+        <param name="lat_key" type="text" value="SiteLat" label="The key of the column for the latitude" optional="False"/>
+        <param name="long_key" type="text" value="SiteLong" label="The key of the column for the longitude" optional="False"/>
+        <param name="tol_spatial_coordinates" type="float" value="0.2" label="A tolerance on the latitude/longitude for the request point" optional="False"/>
+        <param name="depth_key" type="text" value="Depth" label="The key of the column for the observation time" optional="False"/>
+        <param name="date_key" type="text" value="SurveyDate" label="The key of the column for the observation time" optional="False"/>
+        <param name="cmems_username" type="text" value="xxxxx" label="Your login to access CMEMS data"/>
+        <param name="cmems_password" type="text" value="xxxxx" label="Your password to access CMEMS data"/>
```
Comment (bgruening) [inline tools/medenv/med_environ.xml:58]:
```suggestion
```

File: tools/medenv/med_environ.xml [hunk 7]
```diff
@@ -0,0 +1,141 @@
+<tool id="iabiodiv_smartbiodiv_med_environ" name="Adds the environment variables to a dataset" version="0.0.1" profile="22.05">
+    <description>
+        From the input dataset, it will request Copernicus for environment variables.
```
Comment (bgruening) [inline tools/medenv/med_environ.xml:3]: the `name` + `description` will be shown in the toolbox together. The description is a bit cryptic.

maybe something like `name="Copernicus"` and `description="biodiversity environment variables"`

Comment (jeremyfix) [reply tools/medenv/med_environ.xml:3]: Good point ,sorry for the quick and dirty name/description ; It should now be more explicit.

File: tools/medenv/med_environ.xml (hunk 5 above)
Comment (jeremyfix) [reply tools/medenv/med_environ.xml:23]: @Marie59 Did you have time to look at how to fix the question on the credentials or shall I try to do something similar than the @annefou tool ?  If I deal with it, I would probably need to loop with you @yvanlebras .

Comment (Marie59) [reply tools/medenv/med_environ.xml:23]: For now I don't see another way than how Anne did it.

File: tools/medenv/med_environ.xml [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+    </description>
+    <requirements>
+        <requirement type="package" version="2.1.3">pandas</requirement>
+        <requirement type="package" version="4.9.3">lxml</requirement>
+        <requirement type="package" version="1.0.4">medenv</requirement>
+    </requirements>
+    <creator>
+        <person givenName="Jérémy" familyName="Fix" url="http://www.github.com/jeremyfix/" email="jeremy.fix@centralesupelec.fr"/>
+    </creator>
+	<command>
```
Comment (bgruening) [inline tools/medenv/med_environ.xml:13]:
```suggestion
	<command><![CDATA[
```

File: tools/medenv/med_environ.xml [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+    --out_file '${out_file}'
+    --lat_key  '${lat_key}'
+    --long_key '${long_key}'
+    --date_key '${date_key}'
+    --depth_key '${depth_key}'
+    --variables '${env_variables}'
+    --tol_spatial_coordinates '${tol_spatial_coordinates}'
+		--keyfile .medenvkey
+    --verbose '${verbose}'
+    </command>
```
Comment (bgruening) [inline tools/medenv/med_environ.xml:34]:
```suggestion
    ]]></command>
```

File: tools/medenv/med_environ.xml [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+python '$__tool_directory__/med_environ.py'
+    --datafile '${input}'
+    --out_file '${out_file}'
+    --lat_key  '${lat_key}'
+    --long_key '${long_key}'
+    --date_key '${date_key}'
+    --depth_key '${depth_key}'
+    --variables '${env_variables}'
+    --tol_spatial_coordinates '${tol_spatial_coordinates}'
+		--keyfile .medenvkey
```
Comment (bgruening) [inline tools/medenv/med_environ.xml:32]:
```suggestion
    --keyfile .medenvkey
```

**Comment by jeremyfix** [comment]
@bgruening  I validaed them all ; I'm just still struggling with how to introduce the credentials in the User preference menu. 

I tried to debug it using planemo but cannot find the way to add my credentials forms in the User preference yet;

**Comment by yvanlebras** [comment]
THANK you Björn! @annefou can you help us / @jeremyfix / @Marie59 to test the creation of such solution of use of credentials ? For this tool: https://github.com/NordicESMhub/galaxy-tools/blob/master/tools/c3s/c3s.xml#L12 ? Jeremy want to test it on planemo but I don't know how to do so and have not receive detailled procedures to do so....

**Comment by annefou** [comment]
> THANK you Björn! @annefou can you help us / @jeremyfix / @Marie59 to test the creation of such solution of use of credentials ? For this tool: https://github.com/NordicESMhub/galaxy-tools/blob/master/tools/c3s/c3s.xml#L12 ? Jeremy want to test it on planemo but I don't know how to do so and have not receive detailled procedures to do so....

I think I create (manually) a file $HOME/.cdsapirc with my credentials. You will get the "error/warning" message but it should still work.

**Comment by jeremyfix** [comment]
@annefou  I think my question is more about how to make the following work :

```
$__user__.extra_preferences.get('c3s_account|c3s_cds_apikey', "")
```

Which is from the xml file ; I'm not sure how and where this user extra-preference is defined. So far, I was testing locally with planemo and I do not know if I can add custom extra preferences.  I was hopping to test locally on my computer before pushing the modification. 

Anyway,  I just logged in on the https://ecology.usegalaxy.eu , and there I find the c3s_account field. I suppose I should request some admins of https://ecology.usegalaxy.eu to add some   login/password fields for CMEMS in this User/ Manage information page ? @bgruening  @yvanlebras  is that one of you that is expected to make this modification ?

**Comment by yvanlebras** [comment]
Thank you Jeremy! It would be better to test locally before but no clear / direct info on that for now... Aside the creation of the file $HOME/.cdsapirc file @annefou did you make something particular on planemo / local Galaxy server so it can propose the specific credential option in the user tab?

**Comment by annefou** [comment]
> Thank you Jeremy! It would be better to test locally before but no clear / direct info on that for now... Aside the creation of the file $HOME/.cdsapirc file @annefou did you make something particular on planemo / local Galaxy server so it can propose the specific credential option in the user tab?


To test locally with planemo, I use my own .cdsapirc and it works. Have you tried?

**Comment by jeremyfix** [comment]
@yvanlebras  @bgruening  

I changed the code to use credentials from a keyfile.  

The way I was able to test it with my local planemo was to tweak the `med_environ.xml` echoing the login/password in the `~/.medenvkey` file within the <command> tag.

In other words, I was not able to test the code within the <configfile> tag from the xml nor the section

```
		#if $cmems_login == ""
			echo "Error. Set your credentials via: User -> Preferences -> Manage Information"  &&
		#else
			cp '$cmems_credentials' ~/.medenvkey &&
		#end if

```

and finally, for this to work, I suppose you need to add a dedicated section for this tool in the User/Preferences/Manage Information of your galaxy instance.

**Comment by bgruening** [comment]
> and finally, for this to work, I suppose you need to add a dedicated section for this tool in the User/Preferences/Manage Information of your galaxy instance.

This is correct and we have that on EU already. It is independent of the tool that you develop, you just say here what you expect and rely on that. 
Do we know if the same KEY works for this tool and the climate copernicus tool?

**Comment by annefou** [comment]
Each service has its own key / credentials. Users need to register individually to each "data portal" and add their key to Galaxy to be able to retrieve data from Galaxy.

**Comment by yvanlebras** [comment]
THANK YOU so much Jeremy, Björn, Marie, Anne ! Seems to me we are almost ok, but if I well understand, we need to have a dedicated CMEMS oriented credential functionality on usegalaxy.eu as we can't reuse existing copernicus ones. Is it ok for you @bgruening ? How we can do it?

**Comment by yvanlebras** [comment]
THANK YOU Jeremy! I just accepted the CI to run! We will see with Björn for the credential aspect!

**Comment by yvanlebras** [comment]
Is it ok for you @bgruening to merge? Have a nice day !

File: tools/medenv/med_environ.xml [hunk 11]
```diff
@@ -0,0 +1,155 @@
+<tool id="iabiodiv_smartbiodiv_med_environ" name="Adds environment variables collected from copernicus and etopo to a dataset" version="0.0.1" profile="22.05">
+	<description>
+		Given an input dataset which provides geolocalized and timestamped observations, i.e. which possesses columns for the longitude, latitude, depth and time of observation, this tool will augment the dataset with environmental variables as provided by copernicus and etopo. This version of the tool gets environmental data from etopo for the bathymetry and otherwise copernicus mediterranean products.
```
[the remark this answers is not in the document]
Comment (yvanlebras) [reply tools/medenv/med_environ.xml]: True, thank you!

File: tools/medenv/med_environ.xml [hunk 12]
```diff
@@ -0,0 +1,155 @@
+<tool id="iabiodiv_smartbiodiv_med_environ" name="Adds environment variables collected from copernicus and etopo to a dataset" version="0.0.1" profile="22.05">
```
[the remark this answers is not in the document]
Comment (yvanlebras) [reply tools/medenv/med_environ.xml]: yes! Thanks

File: tools/medenv/med_environ.xml [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+        - med-cmcc: https://resources.marine.copernicus.eu/product-detail/MEDSEA_MULTIYEAR_PHY_006_004/INFORMATION
+        - med-ogs: https://resources.marine.copernicus.eu/product-detail/MEDSEA_MULTIYEAR_BGC_006_008/INFORMATION
+
+        The bathymetry is obtained with the ETOPO at : 
+
+        - https://www.ngdc.noaa.gov/mgg/global/relief/ETOPO1/data/bedrock/grid_registered/netcdf/ETOPO1_Bed_g_gmt4.grd.gz
+
+        We request data around the point with a given tolerance and average the collected values. The medenv package does not force that averaging and could return a raster. A later update of the Galaxy tool could take this into account.
+
+        The features you can request are the following, some indexed by depth :
```
[the remark this answers is not in the document]
Comment (yvanlebras) [reply tools/medenv/med_environ.xml:116]: you mean in the "env_variables" param ? It seems to me true but always of interest to have this detail in the help section I think ?

**Comment by yvanlebras** [comment]
Seems to me we are near the end ;) THANK YOU all! @jeremyfix do you think you can see for lasts comments (adding asserts for test, data_column for "depth_key" param, maybe remove `tools/medenv/medenvkey.cfg` file if no more needed and see for `cmems_login` / `cmems_username` naming + help info ?

File: tools/medenv/med_environ.xml [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+    <requirements>
+        <requirement type="package" version="2.1.3">pandas</requirement>
+        <requirement type="package" version="4.9.3">lxml</requirement>
+        <requirement type="package" version="1.0.4">medenv</requirement>
+    </requirements>
+    <creator>
+        <person givenName="Jérémy" familyName="Fix" url="http://www.github.com/jeremyfix/" email="jeremy.fix@centralesupelec.fr"/>
+    </creator>
+	<command><![CDATA[
+		#set $cmems_login = $__user__.extra_preferences.get('cmems_login', "")
```
[the remark this answers is not in the document]
Comment (jeremyfix) [reply tools/medenv/med_environ.xml]: Correct, I fixed this.

File: tools/medenv/med_environ.xml (hunk 13 above)
[the remark this answers is not in the document]
Comment (jeremyfix) [reply tools/medenv/med_environ.xml:116]: I prefer to keep it there to remember that the galaxy wrapper on top of medenv could be improvded to add that degree of freedom.

File: tools/medenv/med_environ.xml [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+            <option value="sea-surface-salinity">sea surface salinity</option>
+            <option value="sea-surface-above-geoid">sea surface above geoid</option>
+            <option value="surface-partial-pressure-co2">surface partial pressure CO2</option>
+            <option value="surface-co2-flux">surface CO2 flux</option>
+            <option value="mixed-layer-thickness">mixed layer thickness</option>
+        </param>
+        <param name="lat_key" type="text" value="SiteLat" label="The key of the column for the latitude" optional="False"/>
+        <param name="long_key" type="text" value="SiteLong" label="The key of the column for the longitude" optional="False"/>
+        <param name="tol_spatial_coordinates" type="float" value="0.2" label="A tolerance on the latitude/longitude for the request point" optional="False"/>
+        <param name="depth_key" type="text" value="Depth" label="The key of the column for the observation time" optional="False"/>
```
[the remark this answers is not in the document]
Comment (jeremyfix) [reply tools/medenv/med_environ.xml]: I tried this , but it does not completely work,; with respect to `use_header_names="true"`

If I let it to the default , which is False I suppose, I correctly get several columns but named as Column 1, Column 2, etc..

If I turn on the use of the header names, I get just one column being detected, with all the names merged :

image

The first lines of the csv I was using as input is 

```
FID,Key,SurveyID,Country,Ecoregion,Realm,SiteCode,Site,SiteLat,SiteLong,SurveyDate,Depth,Phylum,Class,Family,Taxon,Block,Total,Diver,geom
M1_DATA.213611,213611,912342199,Italy,Adriatic Sea,Temperate Northern Atlantic,IT2,La Vela,43.55,13.63,2012-09-22T14:00:00Z,6,Chordata,Actinopterygii,Atherinidae,Atherina hepsetus,2,30,BS,POINT (13.63 43.55)
```

I suppose the use_header_names is working with Tabular inputs ; But is it also expected to work with CSV inputs ?

[the remark this answers is not in the document]
Comment (bgruening) [reply tools/medenv/med_environ.xml]: This could be, can you please fill an upstream issue at https://github.com/galaxyproject/galaxy/issues

Thanks!

**Review by bgruening — APPROVED** [review]

File: tools/medenv/med_environ.xml (hunk 15 above)
[the remark this answers is not in the document]
Comment (jeremyfix) [reply tools/medenv/med_environ.xml]: Done here : https://github.com/galaxyproject/galaxy/issues/17468

**Comment by yvanlebras** [comment]
Woaou! Thank you Jeremy! Seems there is an error with the `use_header_names="true"` as this is number and not column names who are return... I think this is maybe due to the use of a csv file ? Maybe it will work with a test tabular file ?

**Comment by bgruening** [comment]
@jeremyfix can you look at the failing test please?

> parameter 'lat_key': an invalid option ('SiteLat') was selected (valid options: 13,18,6,7,3,16,11,10,2,1,19,4,9,17,5,14,8,12,15,20), parameter 'long_key': an invalid option ('SiteLong') was selected (valid options: 13,18,6,7,3,16,11,10,2,1,19,4,9,17,5,14,8,12,15,20), parameter 'depth_key': an invalid option ('Depth') was selected (valid options: 13,18,6,7,3,16,11,10,2,1,19,4,9,17,5,14,8,12,15,20), parameter 'date_key': an invalid option ('SurveyDate') was selected (valid options: 13,18,6,7,3,16,11,10,2,1,19,4,9,17,5,14,8,12,15,20)

**Comment by jeremyfix** [comment]
> @jeremyfix can you look at the failing test please?
> 
> > parameter 'lat_key': an invalid option ('SiteLat') was selected (valid options: 13,18,6,7,3,16,11,10,2,1,19,4,9,17,5,14,8,12,15,20), parameter 'long_key': an invalid option ('SiteLong') was selected (valid options: 13,18,6,7,3,16,11,10,2,1,19,4,9,17,5,14,8,12,15,20), parameter 'depth_key': an invalid option ('Depth') was selected (valid options: 13,18,6,7,3,16,11,10,2,1,19,4,9,17,5,14,8,12,15,20), parameter 'date_key': an invalid option ('SurveyDate') was selected (valid options: 13,18,6,7,3,16,11,10,2,1,19,4,9,17,5,14,8,12,15,20)

Sure, I'm on it. 

This is due to using data_column . And I will also change my test files to tsv , the input to tabular (this is fixing the issue with `use_header_names`), and I need to adapt the pandas for accessing columns by number not by column names as the data_column seems to return column numbers and not column names.

**Comment by yvanlebras** [comment]
A big big big huge THANK YOU Jérémy!!!!!

File: tools/medenv/med_environ.xml (hunk 15 above)
[the remark this answers is not in the document]
Comment (jeremyfix) [reply tools/medenv/med_environ.xml]: I changed the input format from csv to TSV , This allows detecting header names correctly .

File: tools/medenv/med_environ.xml [hunk 16]
```diff
@@ -0,0 +1,155 @@
...
+            <param name="env_variables" value="bathymetry"/> <!-- This test is requesting etopo only -->
+            <param name="lat_key" value="SiteLat"/>
+            <param name="long_key" value="SiteLong"/>
+            <param name="tol_spatial_coordinates" value="0.2"/>
+            <param name="depth_key" value="Depth"/>
+            <param name="date_key" value="SurveyDate"/>
+            <param name="verbose" value="False"/>
+            <output name="out_file">
+                <assert_contents>
+                    <has_n_lines n="100"/>
```
[the remark this answers is not in the document]
Comment (jeremyfix) [reply tools/medenv/med_environ.xml:86]: I added a check on the number of columns ;

I saw there is a possibility to check for some lines. But I do not know if it is possible, in a test to compare the output tabular  with an expected output ; 

@bgruening  @yvanlebras  do you think this is possible ? 

Note however that this test will still be limited and only check against values obtained from etopo and not cmems, given the required credentials for getting data from copernicus.

[the remark this answers is not in the document]
Comment (bgruening) [reply tools/medenv/med_environ.xml:86]: > I saw there is a possibility to check for some lines. But I do not know if it is possible, in a test to compare the output tabular with an expected output ;

This is the default behavior :)
You find many more options here: https://docs.galaxyproject.org/en/latest/dev/schema.html#tool-tests-test-output

**Comment by bgruening** [comment]
Very cool!

**Comment by jeremyfix** [comment]
I will be happy to give a try once deployed online , especially for testing the credentials set from the user preferences;

Maybe @yvanlebras  , you know where I will be able to test it now that it is merged ?

**Comment by yvanlebras** [comment]
THANK YOU Jeremy!!! AMAZING !!!! And of course, thank you wonder Björn!!! Normally, tools will be automaticaly availbale on https://ecology.usegalaxy.eu next monday, if I correctly add some elements on our CI before sunday ;) For the CMEMS credential, I think Björn you need to create something on eu server isn't it?

**Comment by bgruening** [comment]
Step 3 from https://usegalaxy-eu.github.io/posts/2020/08/22/three-steps-to-galaxify-your-tool/ is missing.

For the credentials, you need to add a similar setting as this https://github.com/usegalaxy-eu/infrastructure-playbook/blob/master/files/galaxy/config/user_preferences_extra_conf.yml#L218 to the same file.

**Comment by yvanlebras** [comment]
Step 3 in progress! 
https://github.com/usegalaxy-eu/usegalaxy-eu-tools/pull/674

**Pull request merged**